File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/factories/InformeFactory.java
package org.example.factories;

import org.example.controllers.DuplaController;
import org.example.models.Dupla;
import org.example.models.Informe;
import org.example.models.PrecipitacionInforme;
import org.example.models.TemperaturaInforme;

import java.time.LocalDate;
import java.util.LinkedList;
import java.util.List;

public class InformeFactory {
    //Singleton
    private static InformeFactory instance = null;
    public static InformeFactory getInstance() {
        if (instance == null) {
            instance = new InformeFactory();
        }
        return instance;
    }

    public Informe createInformeMadrid(DuplaController controller, LocalDate day){
        Dupla max = controller.maxTemMadrid(day);
        Dupla min = controller.minTemMadrid(day);
        Double mediaPrecipitacion = controller.mediaPrecipitacionMadrid(day);
        return new Informe(
                day,
                controller.mediaTemMadrid(),
                new TemperaturaInforme(
                        max.getPoblacion(), max.getTemMax(), max.getTimeMax()
                ),
                new TemperaturaInforme(
                        min.getPoblacion(), min.getTemMax(), min.getTimeMax()
                ),
                new PrecipitacionInforme(
                        mediaPrecipitacion > 0, mediaPrecipitacion
                )
        );
    }

    public List<Informe> createInformesMadrid(DuplaController controller){
        List<Informe> informes = new LinkedList<>();
        informes.add(createInformeMadrid(controller, LocalDate.of(2017, 10, 29)));
        informes.add(createInformeMadrid(controller, LocalDate.of(2017, 10, 30)));
        informes.add(createInformeMadrid(controller, LocalDate.of(2017, 10, 31)));
        return informes;
    }
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/factories/InformeFactory.kt
package factories

import controllers.DuplaController
import models.Informe
import java.time.LocalDate

object InformeFactory {
    fun createInformeMadrid(controller: DuplaController, day: LocalDate): Informe {
        val maxDupla = controller.maxTemMadrid(day)
        val minDupla = controller.minTemMadrid(day)
        val mediaPrecipitacion = controller.mediaPrecipitacionMadrid(day)
        return Informe(
            day,
            controller.mediaTemMadrid(),
            Triple(
                maxDupla.poblacion,
                maxDupla.timeMax,
                maxDupla.temMax
            ),
            Triple(
                minDupla.poblacion,
                minDupla.timeMin,
                minDupla.temMin
            ),
            Pair(
                mediaPrecipitacion != 0.0,
                mediaPrecipitacion
            )
        )
    }
    fun createInformesMadrid(controller: DuplaController): List<Informe> {
        val informes = mutableListOf<Informe>()
        informes.add(createInformeMadrid(controller, LocalDate.of(2017, 10, 29)))
        informes.add(createInformeMadrid(controller, LocalDate.of(2017, 10, 30)))
        informes.add(createInformeMadrid(controller, LocalDate.of(2017, 10, 31)))
        return informes
    }
}